state/immutable-trie/snapshot.go

```diff
@@ -0,0 +1,223 @@
+package itrie
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/dogechain-lab/dogechain/crypto"
+	"github.com/dogechain-lab/dogechain/state"
+	"github.com/dogechain-lab/dogechain/types"
+	"github.com/dogechain-lab/fastrlp"
+)
+
+type Snapshot struct {
+	state StateDB
+	trie  *Trie
+}
+
+func (s *Snapshot) GetStorage(addr types.Address, root types.Hash, rawkey types.Hash) (types.Hash, error) {
+	var (
+		err error
+		ss  state.Snapshot
+	)
+
+	if root == types.EmptyRootHash {
+		ss = s.state.NewSnapshot()
+	} else {
+		ss, err = s.state.NewSnapshotAt(root)
+		if err != nil {
+			return types.Hash{}, err
+		}
+	}
+
+	// tricky downcast, but break out recursion
+	snapshot, ok := ss.(*Snapshot)
+	if !ok {
+		return types.Hash{}, fmt.Errorf("invalid type assertion to Snapshot at %s", root)
+	}
+
+	// slot to hash
+	key := crypto.Keccak256(rawkey.Bytes())
+
+	val, err := snapshot.trie.Get(key, s.state)
+	if err != nil {
+		// something bad happen, should not continue
+		return types.Hash{}, err
+	} else if len(val) == 0 {
+		// not found
+		return types.Hash{}, nil
+	}
+
+	p := &fastrlp.Parser{}
```

Comment:
use `fastrlp.DefaultParserPool`  reuse object ?

Reply:
The variable is on stack, will not escape to heap.

Reply:
ok.